docs/design/datacontracts/contract_csharp_api_design.cs

```diff
@@ -0,0 +1,386 @@
+            {
+                // Do some lookup and allocate an instance of the contract requested
+                //
+                // This lookup can either be reflection based, or we can do it based on a source generator.
+            }
+        }
+
+        [DataContractGlobals("FeatureFlags", 1)]
+        public class FeatureFlags_1
+        {
+            public const int FeatureComInterop = 0;
+        }
+
+        [DataContractGlobals("FeatureFlags", 2)]
+        public class FeatureFlags_2
+        {
+            public const int FeatureComInterop = 1;
+        }
+
+        [DataContractAlgorithm(1)]
+        class MethodTableContract_1 : ContractDefinitions.MethodTableContract, IAlgorithmContract
+        {
+            DataContracts.Target Target;
+            readonly uint ContractVersion;
+            public MethodTableContract_1(DataContracts.Target target, uint contractVersion) { Target = target; ContractVersion = contractVersion; }
+
+            public virtual int DynamicTypeID(TargetPointer methodTablePointer) { return new MethodTable(_target, methodTablePointer).dynamicTypeId; }
+            public virtual int BaseSize(TargetPointer methodTablePointer) { return new MethodTable(_target, methodTablePointer).baseSizeAndFlags & 0x3FFFFFFF; }
+        }
+
+        // This is used for version 2 and 3 of the contract, where the dynamic type id is no longer present, and baseSize has a new limitation in that it can only be a value up to 0x1FFFFFFF in v3
+        [DataContractAlgorithm(2, 3)]
+        class MethodTableContract_2 : ContractDefinitions.MethodTableContract, IAlgorithmContract
+        {
+            DataContracts.Target Target;
+            readonly uint ContractVersion;
+            public MethodTableContract_2(DataContracts.Target target, uint contractVersion) { Target = target; }
+
+            public virtual int DynamicTypeID(TargetPointer methodTablePointer)
+            {
+                throw new NotImplementedException();
+            }
+            public virtual int BaseSize(TargetPointer methodTablePointer)
+            {
+                return new MethodTable(_target, methodTablePointer).baseSizeAndFlags & ((ContractVersion == 3) ? 0x1FFFFFFF : 0x3FFFFFFF);
+            }
+        }
+
+        // We use a source generator to generate the actual runtime properties, and api for working with the fields on this type.
+        //
+        // The source generator would fill in most of the apis, and provide a bunch of properties that give a granular failure model where if a particular field isn't defined, it fails at the access point
+        // This example shows access to a type.
+        [DataContractType]
+        partial struct MethodTable
+        {
+            partial void Get_dynamicTypeId_optional(ref int value);
+            partial void Get_baseSizeAndFlags(ref int value);
+
+            [DataContractLayout(1, 8)]
```

Comment:
I think of the Size property as indicating how much space do we expect between elements in an array of this type. For types that would never show up in arrays (like MethodTable) we could omit it.